File: algorithms/algorithms-cplusplus/leetcode/Question_0059_Spiral_Matrix_II.cc
#include <vector>

using namespace std;

class Solution {
 public:
  vector<vector<int>> generateMatrix(int n) {
    if (n <= 0) return vector<vector<int>>();
    vector<vector<int>> res = vector<vector<int>>(n, vector<int>(n));
    int r1 = 0, r2 = n - 1, c1 = 0, c2 = n - 1, v = 1;
    while (r1 <= r2 && c1 <= c2) {
      for (int i = c1; i <= c2; i++) {
        res[r1][i] = v++;
      }
      for (int i = r1 + 1; i <= r2; i++) {
        res[i][c2] = v++;
      }
      for (int i = c2 - 1; i >= c1; i--) {
        res[r2][i] = v++;
      }
      for (int i = r2 - 1; i > r1; i--) {
        res[i][c1] = v++;
      }
      r1++; r2--; c1++; c2--;
    }
    return res;
  }
};

File: algorithms/algorithms-python/leetcode/Question_0059_Spiral_Matrix_II.py
#!/usr/bin/python
# coding: utf-8

class Solution(object):
    def generateMatrix(self, n):
        """
        :type n: int
        :rtype: List[List[int]]
        """
        if n <= 0:
            return None
        res = [[0] * n for i in range(n)]
        r1, r2, c1, c2, v = 0, n - 1, 0, n - 1, 1
        while r1 <= r2 and c1 <= c2:
            for i in range(c1, c2 + 1):
                res[r1][i] = v
                v += 1
            for i in range(r1 + 1, r2 + 1):
                res[i][c2] = v
                v += 1
            for i in range(c2 - 1, c1 - 1, -1):
                res[r2][i] = v
                v += 1
            for i in range(r2 - 1, r1, -1):
                res[i][c1] = v
                v += 1
            r1, r2, c1, c2 = r1 + 1, r2 - 1, c1 + 1, c2 - 1
        return res
